src/data.js

```diff
@@ -82,78 +70,65 @@ module.exports = {
     const domain = props.domain ? (props.domain.x || props.domain) :
       Scale.getBaseScale(props, "x").domain();
-    const samples = Array.isArray(props.y) ? props.y.length : props.samples;
-    const step = Math.max(...domain) / samples;
-    // return an array of x values spaced across the domain,
-    // include the maximum of the domain
-    return union(
-      lodashRange(Math.min(...domain),
-      Math.max(...domain), step),
-      [Math.max(...domain)]
+    const step = Math.max(...domain) / props.samples;
+    const values = union(
+        lodashRange(Math.min(...domain), Math.max(...domain), step),
+        [Math.max(...domain)]
     );
+    // return data objects for values in {x, y} format
+    return values.map((v) => ({x: v, y: v}));
   },
 
-  returnOrGenerateY(props, x) {
-    if (props.y && typeof props.y === "function") {
-      // if y is a function, apply the function y to to each value of the array x,
-      // and return the results as an array
-      return x.map((datum) => props.y(datum));
-    }
-    // y is either a function or an array, and is never undefined
-    // if it isn't a function, just return it.
-    return props.y;
-  },
-
-  formatData(allData, props) {
-    if (!allData) {
+  formatData(dataset, props, stringMap) {
+    if (!dataset) {
       return [];
     }
-    const stringMap = {
+    stringMap = stringMap || {
       x: this.createStringMap(props, "x"),
       y: this.createStringMap(props, "y")
     };
-    const _formatData = (dataset) => {
-      const cleanedData = this.cleanData(dataset, props);
-      return cleanedData.map((data) => {
-        return merge({}, data, {
-          category: this.determineCategoryIndex(data.x, props.categories),
+    const accessor = {
+      x: this.createAccessor(props.x),
+      y: this.createAccessor(props.y)
+    };
+
+    return this.cleanData(dataset, props)
+      .map((datum) => {
+        const x = accessor.x(datum);
+        const y = accessor.y(datum);
+        return merge({}, datum, {
+          category: this.determineCategoryIndex(x, props.categories),
           // map string data to numeric values, and add names
-          x: typeof data.x === "string" ? stringMap.x[data.x] : data.x,
-          xName: typeof data.x === "string" ? data.x : undefined,
-          y: typeof data.y === "string" ? stringMap.y[data.y] : data.y,
-          yName: typeof data.y === "string" ? data.y : undefined
+          x: typeof x === "string" ? stringMap.x[x] : x,
+          xName: typeof x === "string" ? x : undefined,
+          y: typeof y === "string" ? stringMap.y[y] : y,
+          yName: typeof y === "string" ? y : undefined
         });
       });
+  },
+
+  // For components that take multiple datasets
+  formatDatasets(datasets, props) {
+    // string map must be calculated using all datasets and shared
+    const propsWithDatasets = assign({}, props, {data: datasets});
```

Comment:
In this case, `merge` would try to do a deep merge of `datasets` into `props.data`, potentially messing up complicated nested data, while `assign` just sets the new object's `.data` to `datasets` - ie. it replaces the value instead of merging it. I believe the object spread operator syntax (`{...props, data: datasets}`) would have the same effect as `assign` but per our discussion on chat, we decided to stick with assign because it is slightly more clear and because object spread is still a proposal currently.